### PR TITLE
Disallow empty object names

### LINKED ISSUE #9097: Disallow empty objects names
At the moment, Icinga allows `""` or `null` as object names, i.e. the following config files validate:
```
object HostGroup "" {}
```
```
object HostGroup null {}
```

Both create an objects with the empty string as its name, leading to all kinds of weird situation, for example:
* it can't be queried as a single object using the API
* it can't be specified in the `groups` attribute of a host

So I think the correct behavior would be to simply reject this configuration.

### PR DESCRIPTION
**Pull request by yhabteab**
This PR sets the `required` flag for object names to prevent objects with `""` or `null` names from being created.

fixes #9097

### CODE REVIEW

File: lib/config/expression.cpp [hunk 1]
```diff
@@ -949,16 +949,26 @@ ExpressionResult ObjectExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhi
...
+		value = nameres.GetValue();
+	}
+
+	if (!m_Abstract || !m_DefaultTmpl) {
+		if (value.IsEmpty()) {
+			BOOST_THROW_EXCEPTION(ScriptError("Template/object name must not be empty", m_DebugInfo));
+		}
+	}
+
+	if (!m_DefaultTmpl && !value.IsString()) {
```
Comment (Al2Klimov) [inline lib/config/expression.cpp:967]: Any reason for a different condition here?

Comment (yhabteab) [reply lib/config/expression.cpp:967]: What if a user creates a default template with an integer name and wants to import this manually? Anyway, you only mentioned that the creation of a default template with `template T /* null */` has still to work with this PR.

File: lib/config/expression.cpp [hunk 2]
```diff
@@ -949,16 +949,26 @@ ExpressionResult ObjectExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhi
 	CHECK_RESULT(typeres);
 	Type::Ptr type = typeres.GetValue();
 
-	String name;
+	Value value;
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/config/expression.cpp:952]: Nop, though it reflects better to the new Type.

File: lib/config/expression.cpp [hunk 3]
```diff
@@ -949,16 +949,26 @@ ExpressionResult ObjectExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhi
...
 
-		name = nameres.GetValue();
+		value = nameres.GetValue();
+	}
+
+	if (!m_Abstract || !m_DefaultTmpl) {
+		if (value.IsEmpty()) {
+			BOOST_THROW_EXCEPTION(ScriptError("Template/object name must not be empty", m_DebugInfo));
+		}
+	}
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/config/expression.cpp:965]: Actually not!!

File: lib/config/expression.cpp (hunk 1 above)
Comment (yhabteab) [reply lib/config/expression.cpp:967]: Other templates must have a valid name like any other non-abstract objects.

File: lib/config/expression.cpp [hunk 4]
```diff
@@ -949,16 +949,26 @@ ExpressionResult ObjectExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhi
...
+
+	if (!m_DefaultTmpl && !value.IsString()) {
+		BOOST_THROW_EXCEPTION(ScriptError("Template/object name must be a string", m_DebugInfo));
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/config/expression.cpp:968]: Wait, what does your question refer to? To the order of this condition right? `if (!m_DefaultTmpl && !value.IsString()) `

File: lib/config/expression.cpp (hunk 1 above)
Comment (yhabteab) [reply lib/config/expression.cpp:967]: However, when it is checking for emptyness, it should only be checked for normal objects and non default templates. So that your statement of the referenced issue can be guaranteed.

File: lib/config/expression.cpp [hunk 5]
```diff
@@ -949,16 +949,26 @@ ExpressionResult ObjectExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhi
...
 
 	if (m_Name) {
 		ExpressionResult nameres = m_Name->Evaluate(frame, dhint);
 		CHECK_RESULT(nameres);
 
-		name = nameres.GetValue();
+		value = nameres.GetValue();
+	}
+
+	if (!m_Abstract || !m_DefaultTmpl) {
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/config/expression.cpp:961]: Oh, I see what you're talking about now! Do you want me to strip `!m_Abstract` away from the control structure!

[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/config/expression.cpp:961]: Yes.

File: lib/config/expression.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Al2Klimov) [reply lib/config/expression.cpp:968]: Actually I meant that you first check whether the string is empty and second whether it’s a string.

[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/config/expression.cpp:968]: Actually, the first thing I am checking is whether the name is not empty. Because at this time it is not known that it is string type.

File: lib/config/expression.cpp [hunk 6]
```diff
@@ -958,7 +958,17 @@ ExpressionResult ObjectExpression::DoEvaluate(ScriptFrame& frame, DebugHint *dhi
 		name = nameres.GetValue();
 	}
 
-	return VMOps::NewObject(frame, m_Abstract, type, name, m_Filter, m_Zone,
+	if (!m_DefaultTmpl) {
+		if (name.IsEmpty()) {
+			BOOST_THROW_EXCEPTION(ScriptError("Template/object name must not be empty", m_DebugInfo));
```
Comment (Al2Klimov) [inline lib/config/expression.cpp:963]: Doesn’t `ObjectExpression#m_DebugInfo` highlight the whole object in the error message, not just its name?

Comment (julianbrost) [reply lib/config/expression.cpp:963]: Just highlights the `object Host 42` part, not the block following that defines the object properties. So looks fine for me, especially given that it has to handle `object Host {}` as well.

**Review by julianbrost — COMMENTED** [review]
Interestingly, yet another case of subtle differences between object types. The following config already fails validation:
```
object Host "x" {
    check_command = "dummy"
}

object Service "" {
    host_name = "x"
    check_command = "dummy"
}
```

This happens due to this check:
https://github.com/Icinga/icinga2/blob/32c7f7730db154ba0dff5856a8985d125791c73e/lib/config/configitem.cpp#L223-L227

Is that check obsolete now (there are different code branches that set `name`, so that's not the case necessarily)? Only doing the check there (moving the check out of the `if (nc)`) would mean that this wouldn't have an effect on templates.

**Comment by Al2Klimov** [comment]
Please just make sure

```
object CheckCommand "cc" { }

object Host "h" {
	check_command = "cc"
	vars.x = [ "y" ]
}

apply Service "" for (x in host.vars.x) {
	check_command = "cc"
}
```

keeps working. (At least it does with 3bee85ba8.)

File: lib/base/configobject.ti [hunk 7]
```diff
@@ -57,7 +57,7 @@ private:
 abstract class ConfigObject : ConfigObjectBase < ConfigType
 {
 	[config, no_user_modify] String __name (Name);
-	[config, no_user_modify] String "name" (ShortName) {
+	[config, no_user_modify, required] String "name" (ShortName) {
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply lib/base/configobject.ti:60]: I don't have an example but we are forbidding the user to create an object with an empty name, why should the name still be optional!

**Comment by yhabteab** [comment]
> Is that check obsolete now (there are different code branches that set `name`, so that's not the case necessarily)? Only doing the check there (moving the check out of the `if (nc)`) would mean that this wouldn't have an effect on templates.

What if we would generate the name from the NameComposer after the expression validation? So, the check would be performed after the first try ... catch! We would then not have to check for emptiness.

**Review by julianbrost — CHANGES_REQUESTED** [review]
> What if we would generate the name from the NameComposer after the expression validation? So, the check would be performed after the first try ... catch! We would then not have to check for emptiness.

You mean after the following? https://github.com/Icinga/icinga2/blob/32c7f7730db154ba0dff5856a8985d125791c73e/lib/config/configitem.cpp#L244

The full/composed name is used in the error message if validation fails. Apart from that, I wouldn't update the config of an object after validating it.

> or integer names, after all, why should we allow to create objects with integer names at all, but not be able to import templates contains int names, for example. Because the import expression doesn't even allow the user to specify integer names either

To be honest, I don't see what we gain from this change. The name gets converted to a string ob object creation, so objects names in the active configuration consistently use strings anyways, this is just about if the conversion happens implicitly or you have to do it explicitly.

I mean in general, I'm a fan of strongly typed languages, but implicit conversions happen in many places, so unless we want to make the whole config more type-safe in general, not sure if this helps anything.

By the way, the current state of the current PR still accepts this:

```
object Host "x" {
    check_command = "dummy"
}

apply Service 42 {
    check_command = "dummy"
    assign where true
}
```

At the moment, I'm leaning towards doing just 148f5b8416ded6f3bd816293a9219ddec52dbbb7, this on its own seems to fix (more testing needed, but looking good so far) the empty object name issue, the part that's actually causing issues here.

**Comment by yhabteab** [comment]
> By the way, the current state of the current PR still accepts this:

Do you want me to fix this now as well or are you sticking to the below statement?

> At the moment, I'm leaning towards doing just [148f5b8](https://github.com/Icinga/icinga2/commit/148f5b8416ded6f3bd816293a9219ddec52dbbb7), this on its own seems to fix (more testing needed, but looking good so far) the empty object name issue, the part that's actually causing issues here.

**Comment by julianbrost** [comment]
> or are you sticking to the below statement?
> 
> > At the moment, I'm leaning towards doing just [148f5b8](https://github.com/Icinga/icinga2/commit/148f5b8416ded6f3bd816293a9219ddec52dbbb7), this on its own seems to fix (more testing needed, but looking good so far) the empty object name issue, the part that's actually causing issues here.

That one.

**Review by julianbrost — APPROVED** [review]
